src/aspire/reconstruction/mean.py

```diff
@@ -1,17 +1,262 @@
-class MeanEstimator(Estimator):
+class WeightedVolumesEstimator(Estimator):
+    def __init__(self, weights, *args, **kwargs):
+        """
+        Weighted mean volume estimation.
+
+        This is best considered as an r x r matrix of volumes;
+        each volume is a weighted mean least-squares estimator kernel (MeanEstimator).
+        Convolution with each of these kernels is equivalent
+        to performing a projection/backprojection on a volume,
+        with the appropriate amplitude modifiers and CTF,
+        and also a weighting term;
+        the r^2 volumes are each of pairwise products between the weighting vectors given by the columns of wts.
+
+        Note that this is a non-centered Fourier transform, so the zero frequency is found at index 0.
+
+        Formulas and conventions used for Volume estimation are described in:
+
+        “Structural Variability from Noisy Tomographic Projections.”
+        Andén, Joakim, and Amit Singer.
+        SIAM journal on imaging sciences vol. 11,2 (2018): 1441-1492.
+        doi:10.1137/17M1153509
+
+        "Cryo-EM reconstruction of continuous heterogeneity by Laplacian spectral volumes"
+        Amit Moscovich, Amit Halevi, Joakim Andén and Amit Singer
+        Inverse Problems, Volume 36, Number 2, 2020 IOP Publishing Ltd
+        Special Issue on Cryo-Electron Microscopy and Inverse Problems
+        https://doi.org/10.1088/1361-6420/ab4f55
+
+        :param weights: Matrix of weights, n x r.
+        """
+
+        self.weights = weights
+        self.r = self.weights.shape[1]
+        super().__init__(*args, **kwargs)
+        assert self.src.n == self.weights.shape[0]
+
+    def __getattr__(self, name):
+        if name == "precond_kernel":
+            if self.preconditioner == "circulant":
+                # TODO: Discuss precond plans.
+                # self.precond_kernel = FourierKernelMat(
+                #     1.0 / self.kernel.circularize(), centered=True
+                # )
+                raise NotImplementedError(
+                    "Circulant preconditioner not implemented for WeightedVolumesEstimator."
+                )
+            else:
+                if self.preconditioner.lower() not in (None, "none"):
+                    logger.warning(
+                        f"Preconditioner {self.preconditioner} is not implemented, resetting to default of None."
+                    )
+                self.precond_kernel = None
+            return self.precond_kernel
+
+        else:
+            #  raise AttributeError(name)
+            return super().__getattr__(name)
+
+    def compute_kernel(self):
+        """
+        :return: r x r matrix of volumes, shaped (r, r, 2L, 2L, 2L).
+        """
+
+        _2L = 2 * self.src.L
+        # Note, because we're iteratively summing it is critical we zero this array.
+        kernel = np.zeros((self.r, self.r, _2L, _2L, _2L), dtype=self.dtype)
+        sq_filters_f = np.square(evaluate_src_filters_on_grid(self.src))
```

Comment:
If I'm reading this correctly, we're creating an array of size `(L, L, n)`, which will be quite expensive. May be good to consider how this can be batched effectively (perhaps using sliced sources)? (May also want to consider row-majoring this.)

Reply:
Sure, thanks for the suggestion.

It is not possible to use the sliced source directly, since that doesn't slice the filters and filter indices (instead, they're reset, relying on the prior source to compute the filters known about there).  However, it was easy to add an optional index to `evaluate_src_filters_on_grid`. This method only seems to exist for mean and cov3d codes anyway.

I changed to compute them in the existing batched loops, and otherwise this worked out really clean.

Presently this will do a lot of redundant grid calcs, but we'll be able to cache them away in the future.

Reply:
Great. That's much better.